File: demo/android/engine/sdVewModel/BookSdViewModel.java
package org.noear.ddcat.dao.engine.sdVewModel;

import android.text.TextUtils;

import org.noear.ddcat.dao.engine.DdSource;
import org.noear.ddcat.models.SectionModel;
import org.noear.ddcat.viewModels.ViewModelBase;
import org.noear.sited.ISdViewModel;
import org.noear.sited.SdNode;

import java.util.ArrayList;
import java.util.List;

import noear.snacks.ONode;

/**
 * Created by yuety on 16/9/28.
 */

public class BookSdViewModel extends ViewModelBase implements ISdViewModel {
    public final List<SectionModel> sections = new ArrayList<>();

    public String name;
    public String author;
    public String intro;
    public String logo;
    public String updateTime;
    public boolean isSectionsAsc;//输出的section是不是顺排的

    public String bookUrl;

    public BookSdViewModel(String url){
        this.bookUrl = url;
    }


    @Override
    public void loadByConfig(SdNode config){}

    @Override
    public void loadByJson(SdNode config, String... jsons) {
        if (jsons == null || jsons.length == 0)
            return;

        for (String json : jsons) {
            loadByJsonData(config, json);
        }
    }

    private void loadByJsonData(SdNode config, String json){
        ONode data = ONode.tryLoad(json);

        if(DdSource.isBook(config)) {
            if(TextUtils.isEmpty(name)) {
                name = data.get("name").getString();
                author = data.get("author").getString();
                intro = data.get("intro").getString();
                logo = data.get("logo").getString();
                updateTime = data.get("updateTime").getString();

                isSectionsAsc = data.get("isSectionsAsc").getInt() > 0;//默认为倒排
            }
        }


        ONode sl = data.get("sections").asArray();

        for(ONode n : sl){
            SectionModel sec = new SectionModel();
            sec.name = n.get("name").getString();
            sec.url  = n.get("url").getString();

            sec.orgIndex = total();

            sections.add(sec);

            onAddItem(sec);
        }
    }

    //--------------
    public  void clear() { sections.clear();  }

    public int total()  { return sections.size(); }

    public SectionModel get(int idx)
    {
        if(sections == null)
            return null;

        int len = sections.size();
        if(idx>=len || idx<0)
            return null;
        else
            return sections.get(idx);
    }

    protected void onAddItem(SectionModel sec){

    }
}

File: demo/uwp/engine/sdViewModel/BookSdViewModel.cs
using ddcat.uwp.models;
using ddcat.uwp.viewModels;
using Noear.Snacks;
using org.noear.sited;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ddcat.uwp.dao.engine.sdViewModel {

   public abstract class BookSdViewModel : ViewModelBase, ISdViewModel {
        public ObservableCollection<SectionModel> sections { get; private set; } = new ObservableCollection<SectionModel>();
        
        public String name { get; set; }
        public String author { get; set; }
        public String intro { get; set; }
        public String logo { get; set; }
        public String updateTime { get; set; }

        public bool isSectionsAsc;//输出的section是不是顺排的

        public String bookUrl;

        public BookSdViewModel(String url) {
            this.bookUrl = url;
        }

        public virtual void loadByConfig(SdNode config) { }

        public virtual void loadByJson(SdNode config, params String[] jsons) {
            if (jsons == null || jsons.Length == 0)
                return;

            if (DdSource.isBook(config)) {
                String json = jsons[0]; //不支持多个数据块加载
                ONode data = ONode.tryLoad(json);

                name = data.get("name").getString();
                author = data.get("author").getString();
                intro = data.get("intro").getString();
                logo = data.get("logo").getString();
                updateTime = data.get("updateTime").getString();

                isSectionsAsc = data.get("isSectionsAsc").getInt() > 0;//默认为倒排
            }

            foreach (String json in jsons) //支持多个数据块加载
            {
                ONode data = ONode.tryLoad(json);
                ONode sl = data.get("sections").asArray();

                foreach (ONode n in sl) {
                    SectionModel sec = newSection();
                    sec.name = n.get("name").getString();
                    sec.url = n.get("url").getString();

                    sec.bookName = name;
                    sec.bookUrl = bookUrl;
                    
                    sec.orgIndex = total();


                    sections.Add(sec);

                    onAddItem(sec);
                }
            }
        }

        public virtual void clear() {
            sections.Clear();
        }

        public int total() {
            return sections.Count;
        }

        public SectionModel get(int idx) {
            if (sections == null)
                return null;

            int len = sections.Count;
            if (idx >= len || idx < 0)
                return null;
            else
                return sections[idx];
        }

        protected abstract SectionModel newSection();

        protected virtual void onAddItem(SectionModel sec) {

        }
    }
}
